go/teams/box_audit.go

```diff
@@ -0,0 +1,1098 @@
+// previously failed an audit. When a team does pass an audit, it is removed
+// from the queue.
+type BoxAuditQueue struct {
+	Items   []BoxAuditQueueItem
+	Version Version
+}
+
+var _ Versioned = &BoxAuditQueue{}
+
+func (q *BoxAuditQueue) GetVersion() Version {
+	return q.Version
+}
+
+func NewBoxAuditQueue(version Version) *BoxAuditQueue {
+	return &BoxAuditQueue{
+		Items:   nil,
+		Version: version,
+	}
+}
+
+type BoxAuditQueueItem struct {
+	Ctime      time.Time
+	TeamID     keybase1.TeamID
+	BoxAuditID BoxAuditID
+}
+
+// BoxAuditJail contains TeamIDs that have hit a fatal audit failure or the max
+// number of retryable audit failures. Teams in jail will not be reaudited
+// unless they are explicitly loaded by the fast or slow team loaders.
+type BoxAuditJail struct {
+	TeamIDs map[keybase1.TeamID]bool
+	Version Version
+}
+
+var _ Versioned = &BoxAuditJail{}
+
+func (j *BoxAuditJail) GetVersion() Version {
+	return j.Version
+}
+
+func NewBoxAuditJail(version Version) *BoxAuditJail {
+	return &BoxAuditJail{
+		TeamIDs: make(map[keybase1.TeamID]bool),
+		Version: version,
+	}
+}
+
+func (a *BoxAuditor) shouldAudit(mctx libkb.MetaContext, team Team) (bool, *keybase1.BoxAuditAttemptResult, error) {
+	if team.IsSubteam() {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_SUBTEAM
+		return false, &res, nil
+	}
+	if team.IsOpen() {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_OPENTEAM
+		return false, &res, nil
+	}
+	role, err := team.MemberRole(mctx.Ctx(), mctx.CurrentUserVersion())
+	if err != nil {
+		return false, nil, err
+	}
+	if !role.IsOrAbove(keybase1.TeamRole_WRITER) {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_ROLE
+		return false, &res, nil
+	}
+
+	return true, nil, nil
+}
+
+// loadTeamForBoxAudit loads a team once, but if the client
+// has not yet stored BoxSummaryHashes (due to being an old client)
+// it does a force full reload so it is populated.
+func loadTeamForBoxAudit(mctx libkb.MetaContext, teamID keybase1.TeamID) (*Team, error) {
+	return loadTeamForBoxAuditInner(mctx, teamID, false)
+}
+
+func loadTeamForBoxAuditInner(mctx libkb.MetaContext, teamID keybase1.TeamID, force bool) (team *Team, err error) {
+	defer mctx.TraceTimed("loadTeamForBoxAuditInner", func() error { return err })()
+	arg := keybase1.LoadTeamArg{
+		ID:              teamID,
+		ForceRepoll:     true,
+		Public:          teamID.IsPublic(),
+		ForceFullReload: force,
+	}
+
+	team, err = Load(mctx.Ctx(), mctx.G(), arg)
+	if err != nil {
+		return nil, err
+	}
+	if team == nil {
+		return nil, fmt.Errorf("got nil team from loader")
+	}
+
+	// If the team sigchain state was constructed with support for box summary
+	// hashes, the map will be non-nil but empty. It will only be nil if the
+	// state is cached from a team load before box summary hash support.
+	if team.GetBoxSummaryHashes() == nil {
+		if force {
+			return nil, fmt.Errorf("failed to get a non-nil box summary map after full reload")
```

Comment:
there may be teams that still don't have box summary hashes even after full reload, no?

Reply:
yes, these will fail the audit and will be rotated and retried (at which point they will pass since they now have hashes)

Reply:
ah, nice! that's a feature then, perfect. thanks